File: glmRepeated.R
glmRep <- function(reg, data, by, family=gaussian) {
    # repeat a glm on subsets of a data.frame
    # reg = regression formula. e.g. y ~ x
    # data = data.frame
    # by = string indicating level 1 indicator column name. e.g. 'sub'
    # family = error dist and link function as in glm
    L1.names <- unique(data[[by]])

    coefs <- c()
    for (i in 1:length(L1.names)) {
        ss <- paste0(by, "==L1.names[", i, "]")
        nameDat <- subset(data, eval(parse(text=ss)))
        m <- eval(glm(as.formula(reg), family, nameDat))
        coefs <- rbind(coefs, rbind(array(coefficients(m))))
    }

    coefs <- data.frame(coefs)
    colnames(coefs) <- variable.names(m)
    coefs <- cbind(L1.names, coefs)
    colnames(coefs)[1] <- by
    return(coefs)
}

groupStats <- function(dat) {
#compute summary stats for glmRep output data.frame.     
dat <- dat[sapply(dat, is.numeric)]
nrow <- dim(dat)[2]
out <- data.frame(mean=rep(0,nrow), se=rep(0, nrow), t=rep(0,nrow), p=rep(0,nrow))
rownames(out) <- colnames(dat)

for (i in 1:nrow) {
    out[i, "mean"] <- mean(dat[,i])
    out[i, "se"] <- sd(dat[,i]) / sqrt(length(dat[,i]))
    out[i, "t"] <- out[i, "mean"] / out[i, "se"]
    out[i, "p"] <- t.test(dat[,i])[[3]]
}
return(out)
}

File: glm_repeated.py
import pandas as pd
import statsmodels.formula.api as smf
from scipy.stats import ttest_1samp

def lm_rep(formula, dat, by):
    '''
    repeat a linear regression on subsets of a DataFrame
    formula = regression formula string. e.g. 'y ~ x'
    data = pandas DataFrame
    by = column name string indicating data subset for lm e.g. 'subject'
    '''
    units = dat[by].unique()
    coefs = pd.DataFrame()
    for unit in units:
        sdat = dat.loc[dat[by] == unit]
        res = smf.ols(formula, data = sdat).fit()
        scoefs = res.params
        scoefs.name = unit
        coefs = pd.concat([coefs,scoefs], axis = 1)

    coefs = coefs.T
    return(coefs)

 

def group_stats(dat, dif=0):
    '''compute summary stats for lm_rep output DataFrame'''
    mean = dat.mean()
    mean.name = 'mean'
    sem = dat.std() / pow(len(dat), 0.5)
    sem.name = 'sem'
    t = dat.mean() / (dat.std() / pow(len(dat), 0.5))
    t.name = 't'
    p = ttest_1samp(dat, 0)[1]
    p = pd.Series(p, index = mean.index.values.tolist())
    p.name = 'p'
    out = pd.concat([mean, sem, t, p], axis=1)
    return(out)
